src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs

```diff
@@ -207,52 +237,209 @@ public static FormulaValue Text(EvalVisitor runner, EvalVisitorContext context,
+                          .Replace("\u0006", hasAmPm ? dateTime.ToString("%h", culture) : dateTime.ToString("%H", culture));
+
+            // Minute component
+            format = _minutesDetokenizeRegex.Replace(format, dateTime.ToString("mm", culture))
+                          .Replace("\u000A", dateTime.ToString("%m", culture));
+
+            // Second component
+            format = _secondsDetokenizeRegex.Replace(format, dateTime.ToString("ss", culture))
+                          .Replace("\u0008", dateTime.ToString("%s", culture));
+
+            // Milliseconds component
+            format = _milisecondsDetokenizeRegex.Replace(format, dateTime.ToString("fff", culture))
+                          .Replace("\u000E\u000E", dateTime.ToString("ff", culture))
+                          .Replace("\u000E", dateTime.ToString("%f", culture));
+
+            // AM/PM component
+            format = format.Replace("\u0001", dateTime.ToString("tt", culture))
+                           .Replace("\u0002", dateTime.ToString("%t", culture).ToLower());
+
+            return format;
+        }
+
+        private static string TokenizeDatetimeFormat(string format, CancellationToken cancellationToken)
+        {
+            // Temporary replacements to avoid collisions with upcoming month names, etc.
+            format = _ampmReplaceRegex.Replace(format, "\u0001");
+            format = _apReplaceRegex.Replace(format, "\u0002");
+
+            // Find all "m" chars for minutes, before seconds
+            var match = _minutesBeforeSecondsRegex.Match(format);
+            while (match.Success)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                format = format.Substring(0, match.Index) + "\u000A" + format.Substring(match.Index + 1);
+                match = _minutesBeforeSecondsRegex.Match(format);
+            }
+
+            // Find all "m" chars for minutes, after hours
+            match = _minutesAfterHoursRegex.Match(format);
+            while (match.Success)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var afterHourFormat = format.Substring(match.Index);
+                var minuteAfterHourPosition = _minutesRegex.Match(afterHourFormat);
+                var pos = match.Index + minuteAfterHourPosition.Index;
+
+                format = format.Substring(0, pos) + "\u000A" + format.Substring(pos + 1);
+
+                match = _minutesAfterHoursRegex.Match(format);
+            }
+
+            format = Regex.Replace(format, "[mM]", "\u0003");
+            format = Regex.Replace(format, "[dD]", "\u0004");
+            format = Regex.Replace(format, "[yY]", "\u0005");
+            format = Regex.Replace(format, "[hH]", "\u0006");
+            format = Regex.Replace(format, "[sS]", "\u0008");
+            format = format.Replace('0', '\u000E');
```

Comment:
That's a lot of regex operations. Not sure how much we will gain, but we can optimize it with something like

    var sb = new StringBuilder();
    foreach (var c in format) {
      switch(c) {
        case 'm': case 'M': sb.Append('\u0003'); break;
        case 'd': case 'D': sb.Append('\u0004'); break;
        case 'y': case 'Y': sb.Append('\u0005'); break;
        case 'h': case 'H': sb.Append('\u0006'); break;
        case 's': case 'S': sb.Append('\u0008'); break;
        case '0': sb.Append('\u000E'); break;
        default: sb.Append(c); break;
      }
    }
    format = sb.ToString();